api/v3/Payment.php

```
@@ -0,0 +1,291 @@
+ */
+function civicrm_api3_payment_get($params) {
+  $ft = array();
+  if (CRM_Utils_Array::value('contribution_id', $params)) {
+    $params['entity_id'] = $params['contribution_id'];
+  }
+  $params['entity_table'] = 'civicrm_contribution';
+  $eft = civicrm_api3('EntityFinancialTrxn', 'get', $params);
+  if (!empty($eft['values'])) {
+    foreach ($eft['values'] as $efts) {
+      $eftids[] = $efts['financial_trxn_id'];
+      $map[$efts['financial_trxn_id']] = $efts['entity_id'];
+    }
+    $ftParams = array(
+      'id' => array( 'IN' => $eftids ),
+      'is_payment' => 1,
+    );
+    $ft = civicrm_api3('FinancialTrxn', 'get', $ftParams);
+    foreach ($ft['values'] as &$values) {
+      $values['contribution_id'] = $map[$values['id']];
+    }
+  }
+  return civicrm_api3_create_success($ft['values'], $params, 'Payment', 'cancel');
+}
+
+/**
+ * Delete a payment.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_delete(&$params) {
+  $params['id'] = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $params['id'], 'id')) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to delete a payment'));
+  }
+  return civicrm_api3('FinancialTrxn', 'delete', $params);
+}
+
+/**
+ * Cancel/Refund a payment for a Contribution.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_cancel(&$params) {
+  $ftId = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'id')) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to refund a payment'));
```

note that 
1) this should throw an exception &
2) if this will fail somewhere else anyway then don't check for it. Our policy is avoid any pre-save validation in the api if possible. Note that if an sql statement fails there is some post-save validation